### ISSUE TITLE
Onedrive client exits on uploading actions, exception code 141 (but downloading works).

### ISSUE THREAD

**Issue by xmoiduts** [abraunegg/onedrive#1526]
**Note:** Before submitting a bug report, please ensure you are running the latest 'onedrive' client as built from 'master' and by using the latest available DMD compiler. Refer to the readme on building the client for your system. [x]

### Bug Report Details ###
**Describe the bug**
Onedrive client crashes on attempting to upload file/folder, returns exception code 141, but no problems in downloading.

Hi, I have a daily-built server that syncs data to onedrive every day. By accident, I found the server unable to **upload** its generated data to onedrive. The observation of its activity shows that:

- If onedrive cloud has newer (newly created/modified) files than the server (code, config files... sorry that I use onedrive to sync my source code/executables), the server will sync (download) them normally, no problem here.
- If my server has newer file than the onedrive cloud (images, logs,...), then it crashes on(before) uploading this file/directory, nothing will be uploaded, the `--verbose` log ends on 'The file content changes', 'Uploading modified file [filename]', then crashes, `echo $?` will print `141`.

**How I use this OneDrive**
In my use case, full selective sync (sync list) will take too long ('processing 207 205 199 ... files' v.s. 400000+ files), so I use `--single-directory` to speed up the sync process.
Performing full selective sync won't upload any file either;
Performing  '--resync' --single-directory **will upload *some* of newer files**, multiple attempts will be needed to sync all newer files as random uploading crashes exist (not this issue, this issue is not random but 100% triggered). But anyway it's not the right solution, as it takes days now and will be longer afterwards.

**Application and Operating System Details:**
*   Provide your OS & version (CentOS 6.x, Ubuntu 18.x etc) and the output of: `uname -a` `Ubuntu 20.04.2 LTS` and `Linux kedama-crawler 5.4.0-73-generic #82-Ubuntu SMP Wed Apr 14 17:39:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
*   Are you using a headless system (no gui) or with a gui installed? `headless`
*   OneDrive Account Type `Personal`
*   Did you build from source or install from a package? `Build from source`
*   If you installed from source, what is your DMD or LDC compiler version: `dmd --version` or `ldmd2 --version` `v2.097.0`
*   OneDrive Application Version: Output of `onedrive --version` `onedrive v2.4.12-8-g1d4a206`
*   OneDrive Application Configuration: Output of `onedrive --display-config` 
    ```
    Configuration file successfully loaded                                       
    onedrive version                       = v2.4.12-8-g1d4a206                  
    Config path                            = /root/.config/onedrive              
    Config file found in config path       = true                                
    Config option 'check_nosync'           = false                               
    Config option 'sync_dir'               = /mnt/volume_fra1_02/OneDriveMnt/    
    Config option 'skip_dir'               =                                     
    Config option 'skip_file'              = ~*|.~*|*.tmp                        
    Config option 'skip_dotfiles'          = false                               
    Config option 'skip_symlinks'          = false                               
    Config option 'monitor_interval'       = 300                                 
    Config option 'min_notify_changes'     = 5                                   
    Config option 'log_dir'                = /var/log/onedrive/                  
    Config option 'classify_as_big_delete' = 1000                                
    Config option 'upload_only'            = false                               
    Config option 'no_remote_delete'       = false                               
    Config option 'remove_source_files'    = false                               
    Config option 'sync_root_files'        = false                               
    Selective sync 'sync_list' configured  = true                                
    sync_list contents:                                                          
    Codes/kedamadiff-proj/KedamaDiff/                                            
    !Codes/kedamadiff-proj/KedamaDiff-Dev/                                       
    !Codes/kedamadiff-proj/KedamaDiff/.git                                       
    !Codes/kedamadiff-proj/KedamaDiff/.vscode                                    
    !Codes/kedamadiff-proj/KedamaDiff/__pycache__                                
    Codes/kedamadiff-proj/data-production/                                       
    Business Shared Folders configured     = false                               
    ```
*   Provide the version of curl you are using: Output of `curl --version` `curl 7.68.0`
*   Is your configured 'sync_dir' a local directory or a network mount point? `Local-directory'...... kind of, it's a virtual block storage in the cloud datacenter.
*   If *not* local, provide all the mountpoints in your system: Output of: `mount`
*   What partition format type does your configured 'sync_dir' reside on? Output of: `lsblk -f` `sdb`
    ```
    NAME    FSTYPE   LABEL           UUID                                 FSAVAIL FSUSE% MOUNTPOINT             
    loop0   squashfs                                                            0   100% /snap/core18/2066      
    loop1   squashfs                                                            0   100% /snap/lxd/20326        
    loop2   squashfs                                                            0   100% /snap/snapd/11841      
    loop3   squashfs                                                            0   100% /snap/snapd/12057      
    --> sdb     ext4                     55e109f3-a5e8-48f1-bf84-a311c4d5f56c      6G    75% /mnt/volume_fra1_02    
    vda                                                                                                         
    ├─vda1  ext4     cloudimg-rootfs e17a595e-18f3-45e3-8f22-e0d1005fbae1   17.3G    28% /                      
    ├─vda14                                                                                                     
    └─vda15 vfat     UEFI            639E-09CF                              96.6M     7% /boot/efi              
    ```
*   Explain your entire configuration setup - is the OneDrive folder shared with any other system, shared with any other platform at the same time, is the OneDrive account you use shared across multiple systems / platforms / Operating Systems and in use at the same time

    `Shared with my PC, but only one of the PC/server will write to this folder at the same time.`

**Note:** Please generate a full debug log whilst reproducing the issue as per [https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support](https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support) and email to support@mynas.com.au [x]

**To Reproduce**
Steps to reproduce the behavior if not causing an application crash:
[May be hard to reproduce]
onedrive --synchronize --verbose --local-first --single-directory <sync dir> # this will fail
onedrive --synchronize --verbose --local-first --single-directory <sync dir> --resync # this will anyway manage to upload some files, given that the server is the trusted source of different files.

If issue is replicated by a specific 'file' or 'path' please archive the file and path tree & email to support@mynas.com.au 

**Complete Verbose Log Output**
A clear and full log of the problem when running the application in the following manner (ie, not in monitor mode):
```bash
onedrive --synchronize --verbose <any of your other needed options>
```

<too long, see screenshots>

Run the application in a separate terminal window or SSH session and provide the entire application output including the error & crash. When posing the logs, Please format log output to make it easier to read. See [https://guides.github.com/features/mastering-markdown/](https://guides.github.com/features/mastering-markdown/) for more details.

Application Log Output:
```bash
Verbose console log output goes here
```

<too long, see screenshots>

**Screenshots**
![image](https://user-images.githubusercontent.com/16047865/121934008-57a37880-cd79-11eb-84e4-e72f5f07d394.png)
![image](https://user-images.githubusercontent.com/16047865/121934096-7144c000-cd79-11eb-98e7-ea63d63249db.png)
![image](https://user-images.githubusercontent.com/16047865/123509395-d23c9400-d6a7-11eb-846f-1a698fd5ce29.png)

**Additional context**
Several issues mentioned cURL bugs, but in this issue, the onedrive sometimes succeeds to upload (in resync mode), and downloading works all the time. So I don't think cURL is the issue. 

When syncing/resyncing with --single-directory, the upload issue seems to happen a lot, but I can't confirm if this flag is bound to upload failure as a full sync(not resync) will also fail to upload. In some attempts, --resync + no --single-directory would succeed re-syncing.

### Bug Report Checklist ###
*   [x] Detailed description
*   [x] Application and Operating System Details provided in full
*   [] Reproduction steps (if applicable)
*   [] Verbose Log Output from your error
*   [x] Debug Log generated and submitted

**Comment by abraunegg** [comment]
@xmoiduts 
This 99% sounds like this: https://github.com/abraunegg/onedrive/blob/master/docs/known-issues.md#application-stops-running-without-any-visible-reason

A shell exit code of `141` will generally indicate a broken pipe or broken TCP connection.

Please can you perform the HTTPS Debug logging to analyse what is going on when this is occurring for you.

**Comment by xmoiduts** [comment]
Okay, I'll debug https later after finishing my full resync, while I still intend to emphasize that, in a sync attempt, the client compares 100000s of files(which may introduce a lot of internet requests), no problem here, until the client makes its first upload request.

**Comment by abraunegg** [comment]
@xmoiduts 
> after finishing my full resync

Please do *not* use `--resync` thinking this will solve your issue. It will not solve your issue - it probably will exacerbate the problem you are seeing.

What `--resync` does is it removes the local cache database, thus *forces* the client to go through *all* your files online and local to re-populate the cache database. This is *very* time intensive. You should *only* ever use `--resync` if the application tells you to after some major configuration change.

> will take too long ('processing 207 205 199 ... files' v.s. 400000+ files)
> the client compares 100000s of files

OneDrive supports a maximum of 300K files - if you have more than this, you are going to run into OneDrive API issues. This is regardless of OneDrive account type - Personal, Office365, SharePoint etc

Combined with you using `--resync` these two factors are what is impacting you.

The most likely scenario here is, the TCP connection out to OneDrive for the original 'setup' when the application starts is not being refreshed because of having to go through your file volume levels, and because of you using `--resync` you are making this worse for yourself. This is because `--resync` throws away also the 'checkpoint' URL where only the latest changes from online are being checked.

Because of this scenario, and potentially some sort if Internet device (you Linux box, your Internet router, ISP or something else in-between) is not honoring the TCP connection state (which should be still valid based on the curl options that are used) and is tearing down the connection. The HTTPS Debug will give you this additional insight what is breaking, but will not tell you where unfortunately. 

In short:
* Dont have more than 300K files on OneDrive
* Dont use `--resync` unless the application tells you to

**Comment by abraunegg** [comment]
@xmoiduts
Any update here regarding your investigations?

**Comment by xmoiduts** [comment]
Hi, I debugged https, its log says a connection #1 lasted 675s became too old and was disconnected. (using single-directory and not resync).
I understand your explaination of flakky internet. But in my previous opinion it should be far more than 1 connection during the one-time sync procedure even if in --single-directory mode, doesn't it make new internet request every 200-ish files?

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by abraunegg** [comment]
@xmoiduts 
> But in my previous opinion it should be far more than 1 connection during the one-time sync procedure even if in --single-directory mode, doesn't it make new internet request every 200-ish files?

No it does not create a new connection. 

The first connection is created when the application starts up. This connection is then 'reused' to ensure that the TCP aged time does not time out, and the connection to the OneDrive service is maintained through the entire application running process.

When you see:
```
processing 207 ...
processing 205 ...
```

OneDrive is sending 'changes' or 'items' from online to the client as per the API specification in ~200 items at a time. These items are actually JSON items, contained in a single JSON response. A new connection is not created for each new ~200 set of items. At the end of the current bundle, a new request, using the 'netxtLink' URL is sent to OneDrive using existing established connection.

> connection #1 lasted 675s became too old and was disconnected

So there could be a couple of things here:
1. Curl itself is the one failing. We have a timer in Curl that *is* set by the application to 600 seconds which relates to the timeout for activity on connection and dealing with Curl's CURLOPT_LOW_SPEED_TIME option. If this is exceeded, Curl will consider this to be too slow and abort. This *could* be what you are seeing. Generating a PR to test a fix is simple.
2. You have some sort of device - local Linux platform TCP options, your Internet router, your ISP, *something else* between you and OneDrive, where it considers 600 seconds for a TCP connection not doing anything, and shutting down your connection. This is something I cannot solve for or fix

**Comment by abraunegg** [comment]
@xmoiduts 

Please can you test the following PR to see if this resolves this issue:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/1534/head:pr1534
git checkout pr1534
./configure; make clean; make;
```
To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

When running the PR, your version should be: `onedrive v2.4.12-12-g55c917d` or greater

**Comment by xmoiduts** [comment]
Unfortunately it didn't resolve the issue, even though in the PR I saw the
timeout being changed from 600s to 1800s, the curl request still reported a
failure and crash after 600-ish seconds. Here I will attach two logs:
one(log 5) for 2.4.12.8-g1d4a206 and the other(log6) for 2.4.12.12-g55c917d.
Though they look very similar, their onedrive versions are different.

BTW, as the log says 'connection #1 is dead, disconnect it', does it mean
there are potential to re-initialize connection #2, #3... after such
failure?

I directly installed the g55c917d version, overwriting the old one,
`onedrive --version` confirms the installation success.

abraunegg ***@***.***> 于2021年6月30日周三 上午6:03写道：

**Comment by xmoiduts** [comment]
Unfortunately it didn't resolve the issue, even though in the PR I saw the
timeout being changed from 600s to 1800s, the curl request still reported a
failure and crash after 600-ish seconds. Here I will attach two logs:
one(log 5) for 2.4.12.8-g1d4a206 and the other(log6) for 2.4.12.12-g55c917d.
[sent separately] Though they look very similar, their onedrive versions
are different.

BTW, as the log says 'connection #1 is dead, disconnect it', does it mean
there are potential to re-initialize connection #2, #3... after such
failure?

I directly installed the g55c917d version, overwriting the old one,
`onedrive --version` confirms the installation success.

abraunegg ***@***.***> 于2021年6月30日周三 上午6:03写道：

**Comment by abraunegg** [comment]
@xmoiduts 
As I indicated - this 'may' have been the issue, and it looks like the PR did not resolve this.

Actions for you - this is an issue you need to solve:
* Dont use Ubuntu, install another Linux OS that is not Ubuntu or Ubuntu based and test using that
  * Dont run the other OS as a VM under Ubuntu - as, you are still using the Ubuntu network stack - build a totally new PC
  * If you still experience the issue on the totally different non-Ubuntu OS, then it is not Ubuntu related .... if you dont encounter the issue - go raise this issue with Ubuntu .. or stop using Ubuntu
* Debug your network and your connection to OneDrive
  * Just because 'other' things work, does not mean you dont have an issue somewhere that other applications do not uncover
* Raise a trouble ticket with the Curl team and get them to assist you.
  * This client relies on the Curl Library - it passes all the 'work' actions to Curl. Your debug output shows that it is Curl that is terminating the action - not this client
* Complain to your ISP about connection drops to OneDrive
* Open an issue ticket with Microsoft

**Comment by abraunegg** [comment]
Closing issue for the following reasons:
* Not a bug
* Something environmental with users setup - Linux host or something in-between that and OneDrive
